File: inputs/kernel0/kernel0.c
// preprocessing
#define	EBUSY		16



// struct/typedef definitions.
typedef int size_t;

typedef long long	__kernel_loff_t;
typedef __kernel_loff_t		loff_t;

struct ctl_table {
	void *data;
	int maxlen;
	void *extra1;
	void *extra2;
};

typedef struct {
	int counter;
} atomic_t;
typedef atomic_t atomic_long_t;

typedef struct {
	unsigned int lock;
} arch_spinlock_t;

typedef struct raw_spinlock {
	arch_spinlock_t raw_lock;
} raw_spinlock_t;

struct list_head {
	struct list_head *next, *prev;
};

struct mutex {
	atomic_long_t		owner;
	raw_spinlock_t		wait_lock;
	struct list_head	wait_list;
};





// function prototypes
int proc_dointvec_minmax(struct ctl_table *table, int write, void *buffer, size_t *len, loff_t *pos);
void mutex_lock(struct mutex *lock);
int atomic_read(const atomic_t *v);
void mutex_unlock(struct mutex *lock);


// global variables
atomic_t nr_callchain_events;
struct mutex callchain_mutex;


// function to check, from callchain.c
int perf_event_max_stack_handler(struct ctl_table *table, int write, void *buffer, size_t *lenp, loff_t *ppos) { 
	int *value = (*table).data;
	int new_value = *value, ret;
	struct ctl_table new_table = *table;

	new_table.data = &new_value;
	ret = proc_dointvec_minmax(&new_table, write, buffer, lenp, ppos);
	if (ret || !write)
		return ret;

	mutex_lock(&callchain_mutex);
	if (atomic_read(&nr_callchain_events))
		ret = -EBUSY;
	else
		*value = new_value;

	mutex_unlock(&callchain_mutex);

	return ret;
}

File: inputs/kernel0/kernel0.rs
// preprocessing
const EBUSY: i32 = 16;

// struct/typedef definitions.
type size_t = i32;

type __kernel_loff_t = i64;
type loff_t = __kernel_loff_t;

struct ctl_table<'a> {
    data: &'a mut i32,
}

struct atomic_t {
    counter: i32,
}
type atomic_long_t = atomic_t;

struct arch_spinlock_t {
    lock: u32,
}

struct raw_spinlock_t {
    raw_lock: arch_spinlock_t,
}

struct list_head<'a> {
    next: &'a mut list_head<'a>,
    prev: &'a mut list_head<'a>,
}

struct mutex<'a> {
    owner: atomic_long_t,
    wait_lock: raw_spinlock_t,
    wait_list: list_head<'a>,
}

// function prototypes
fn proc_dointvec_minmax(
    table: &mut ctl_table,
    write: i32,
    buffer: &mut i32,
    len: &mut size_t,
    pos: &mut loff_t,
) -> i32 {
    return 0;
}
fn mutex_lock(lock: &mut mutex) {}
fn atomic_read(v: &mut atomic_t) -> i32 {
    return 0;
}
fn mutex_unlock(lock: &mut mutex) {}

// global variables
static nr_callchain_events: atomic_t = atomic_t { counter: 0 };

static mutex_owner: atomic_long_t = atomic_long_t { counter: 0 };
static mutex_wait_lock: raw_spinlock_t = raw_spinlock_t {
    raw_lock: { arch_spinlock_t { lock: 0 } },
};
static mutex_wait_list: list_head = list_head {
    next: &mut mutex_wait_list,
    prev: &mut mutex_wait_list,
};
static callchain_mutex: mutex = mutex {
    owner: mutex_owner,
    wait_lock: mutex_wait_lock,
    wait_list: mutex_wait_list,
};

// function to check, from callchain.c
fn perf_event_max_stack_handler(
    table: &mut ctl_table,
    write: i32,
    buffer: &mut i32,
    lenp: &mut size_t,
    ppos: &mut loff_t,
) -> i32 {
    let mut value: &mut i32 = (*table).data;
    let mut new_value: i32 = *value;
    let mut ret: i32;
    let mut new_table: ctl_table = *table; // Cannout move out of *table behind mutable reference

    new_table.data = &mut new_value;
    ret = proc_dointvec_minmax(&mut new_table, write, buffer, lenp, ppos);
    if ret != 0 || write == 0 {
        return ret;
    }

    mutex_lock(&mut callchain_mutex);
    if atomic_read(&mut nr_callchain_events) != 0 {
        ret = -EBUSY;
    } else {
        *value = new_value;
    }
    mutex_unlock(&mut callchain_mutex);

    return ret;
}

fn main() {
    println!("Runs!");
}
